### PR TITLE
feat(worker/ens): support ens transactions

### LINKED ISSUE #78: ens
### Tag

Collectible

### Project URL

https://ens.domains/

### Worker description

ENS Protocol, Build web3 naming service.

### What actions should be included?

1. Name Registered (V1, V2)
2. Name Renewed
3. TextChanged (w/ w/o value)
4. Name Wrapped & Unwrapped
5. Fuses Set
6. Address Changed
7. Contenthash Changed
8. Name Changed
9. Pubkey Changed

### Additional description

#79 

### This is not a duplicated worker request

- [x] I have searched [existing issues](https://github.com/NaturalSelectionLabs/RSS3-Node/issues) and [pull requests](https://github.com/NaturalSelectionLabs/RSS3-Node/pulls) to ensure this worker proposal has not already been requested

### PR DESCRIPTION
**Pull request by Candinya**
resolve #78

### CODE REVIEW

File: internal/database/dialer/cockroachdb/migration/20240115040941_add_ens_namehash.sql [hunk 1]
```diff
@@ -0,0 +1,18 @@
+-- +goose Up
+-- +goose StatementBegin
+CREATE TABLE IF NOT EXISTS "dataset_ens_namehashes"
+(
+    "hash_hex" text NOT NULL,
```
Comment (brucexc) [inline internal/database/dialer/cockroachdb/migration/20240115040941_add_ens_namehash.sql:5]: To store hex, you can use bytea to store it without converting it to string.

File: internal/engine/worker/contract/ens/utils.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+	"go.uber.org/zap"
+)
+
+// getEnsName get ens name by name hash (node)
+func (w *worker) getEnsName(ctx context.Context, _ *big.Int, namehash common.Hash) (string, error) {
+	// Load name from dataset
+	namehashRecord, err := w.databaseClient.LoadDatasetENSNamehash(ctx, namehash)
+	if err != nil {
```
Comment (brucexc) [inline internal/engine/worker/contract/ens/utils.go:17]:
```bash
if err != nil {
     zap.L().Error("Fail to find ens namehash from dataset", zap.String("namehash", namehash.Hex()), zap.Error(err))

    return "", fmt.Errorf("fail to find ens namehash %v, all methods failed", namehash.String())
}
	
return namehashRecord.Name, nil
```

File: internal/engine/worker/contract/ens/worker.go [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+
+type worker struct {
+	// Base
+	config         *config.Module
+	ethereumClient ethereum.Client
+	tokenClient    token.Client
+	databaseClient database.Client
+	// Specific filters
+	baseRegistrarImplementationFilterer *ens.BaseRegistrarImplementationFilterer
+	ethRegistrarControllerV1Filterer    *ens.ETHRegistrarControllerV1Filterer
```
Comment (brucexc) [inline internal/engine/worker/contract/ens/worker.go:43]: Ambiguous naming, ethRegistrarControllerV1Filterer or ensRegistrarControllerV1Filterer?

Comment (Candinya) [reply internal/engine/worker/contract/ens/worker.go:43]: It's called `ENS: ETH Registrar Controller` in etherscan tag (or `Old` for V1), so it should be fine.

File: internal/database/dialer/cockroachdb/client.go [hunk 4]
```diff
@@ -285,6 +286,38 @@ func (c *client) SaveDatasetFarcasterProfile(ctx context.Context, profile *model
 	return c.database.WithContext(ctx).Clauses(clauses...).Create(&value).Error
 }
 
+// LoadDatasetENSNamehash accepts an ENS namehash, and response with a record containing its original name string
+func (c *client) LoadDatasetENSNamehash(ctx context.Context, hash common.Hash) (*model.ENSNamehash, error) {
+	var value table.DatasetENSNamehash
+
+	if err := c.database.WithContext(ctx).
+		Where("hash = ?", hash.Bytes()).
```
Comment (brucexc) [inline internal/database/dialer/cockroachdb/client.go:294]: use hash directly, no need to convert to bytes

**Review by brucexc — DISMISSED** [review]

**Review by pseudoyu — APPROVED** [review]

**Review by pseudoyu — APPROVED** [review]

**Pull request merged**